### PR TITLE
aws: use existing security groups for masters/workers

### LINKED ISSUE #1163: AWS: Allow Customization of Security Group Ingress/Egress
## Is this a BUG REPORT or FEATURE REQUEST?

FEATURE REQUEST

## Versions

* Tectonic version: [1.6.4-tectonic.1](https://github.com/coreos/tectonic-installer/tree/1.6.4-tectonic.1)
* Terraform version: `0.9.4`
* Platform (aws|azure|openstack|metal): aws

## What happened?

Security Groups created via the [vpc module](https://github.com/coreos/tectonic-installer/tree/1.6.4-tectonic.1/modules/aws/vpc) hard code egress & ingress cidr blocks to `0.0.0.0/0`. Example:

```
resource "aws_security_group_rule" "worker_ingress_ssh" {
  type              = "ingress"
  security_group_id = "${aws_security_group.worker.id}"

  protocol    = "tcp"
  cidr_blocks = ["0.0.0.0/0"]
  from_port   = 22
  to_port     = 22
```

## What you expected to happen?

Optimally, similar behavior used to provide additional tags ([`variable "tectonic_aws_extra_tags` & `variable "tectonic_autoscaling_group_extra_tags"`](https://github.com/coreos/tectonic-installer/blob/1.6.4-tectonic.1/platforms/aws/variables.tf)) could be used to provide either security group ID's (preferred), using the `source_security_group_id` argument ([example](https://github.com/coreos/tectonic-installer/blob/1.6.4-tectonic.1/modules/aws/vpc/sg-worker.tf#L71-L79)) or an input cidr (i.e. `10.0.0.0/16`) to allow communication to/from. Separating Ingress from Egress, instead of one blanket argument, would likely be necessary to allow customization of one versus the other.

## How to reproduce it (as minimally and precisely as possible)?

n/a

## Anything else we need to know?

n/a

### PR DESCRIPTION
**Pull request by hhoover**
This change allows a user to specify custom security groups for
master and worker nodes, which need to exist in advance before
install. If a custom security group is not provided, the groups
will be created. This also assumes you are using an existing VPC.

This is based on the Azure work for custom NSG and makes heavy
use of Terraform's interpolation. This change also preserves the
use of `tectonic_aws_master_extra_sg_ids` and
`tectonic_aws_worker_extra_sg_ids` for users looking to use more
than one security group.

Closes #1163

### CODE REVIEW

File: modules/aws/vpc/outputs.tf [hunk 1]
```diff
@@ -17,11 +17,13 @@ output "etcd_sg_id" {
 }
 
 output "master_sg_id" {
-  value = "${aws_security_group.master.id}"
+  //value = "${aws_security_group.master.id ? aws_security_group.master.id : var.external_sg_master}"
```
Comment (s-urbaniak) [inline modules/aws/vpc/outputs.tf]: do you mind to remove this commented code?

File: modules/aws/vpc/outputs.tf [hunk 2]
```diff
@@ -17,11 +17,13 @@ output "etcd_sg_id" {
...
+  value = "${var.external_sg_master == "" ? aws_security_group.master.id : var.external_sg_master}"
 }
 
 output "worker_sg_id" {
-  value = "${aws_security_group.worker.id}"
+  //value = "${aws_security_group.worker.id ? aws_security_group.worker.id : var.external_sg_worker}"
```
Comment (s-urbaniak) [inline modules/aws/vpc/outputs.tf]: see above

File: modules/aws/vpc/sg-etcd.tf [hunk 3]
```diff
@@ -48,3 +44,43 @@ resource "aws_security_group" "etcd" {
...
+}
+
+resource "aws_security_group_rule" "etcd_ingress_ssh_from_master" {
+  count                    = "${var.external_sg_master == "" && var.enable_etcd_sg ? 1 : 0}"
+  type                     = "ingress"
+  protocol                 = "tcp"
+  from_port                = 22
+  to_port                  = 22
+  security_group_id        = "${aws_security_group.etcd.id}"
+  source_security_group_id = "${aws_security_group.master.id}"
```
Comment (s-urbaniak) [inline modules/aws/vpc/sg-etcd.tf]: can't we unify `etcd_ingress_ssh_from_external_master` and `etcd_ingress_ssh_from_master` doing the following?
```
source_security_group_id = "${var.external_sg_master == "" : aws_security_group.master.id ? var.external_sg_master}"
```

File: modules/aws/vpc/sg-etcd.tf [hunk 4]
```diff
@@ -48,3 +44,43 @@ resource "aws_security_group" "etcd" {
...
+  count                    = "${var.external_sg_master != "" && var.enable_etcd_sg ? 1 : 0}"
+  type                     = "ingress"
+  protocol                 = "tcp"
+  from_port                = 2379
+  to_port                  = 2379
+  security_group_id        = "${aws_security_group.etcd.id}"
+  source_security_group_id = "${var.external_sg_master}"
+}
+
+resource "aws_security_group_rule" "etcd_ingress_client_from_master" {
```
Comment (s-urbaniak) [inline modules/aws/vpc/sg-etcd.tf]: see question above

File: modules/aws/vpc/sg-etcd.tf (hunk 3 above)
Comment (hhoover) [inline modules/aws/vpc/sg-etcd.tf]: The resource is computed before any conditionals, and terraform will throw an error that `aws_security_group.master.id` does not exist when using an external security group.

File: modules/aws/vpc/outputs.tf (hunk 1 above)
Comment (hhoover) [inline modules/aws/vpc/outputs.tf]: done. derp

File: modules/aws/vpc/sg-etcd.tf (hunk 3 above)
Comment (metral) [inline modules/aws/vpc/sg-etcd.tf]: You can wrap `aws_security_group.master.id` in a TF interpolation `join()` to get around the fact that it is conditionally created and implement just one rule as @s-urbaniak is proposing.

See https://github.com/coreos/tectonic-installer/commit/e0c2d67a559aca2d0514c537985b9ce7c0072304#diff-95b96d483bc6fb7c40b02ec5c06c86a9R6 and the commit message for more info.

Comment (hhoover) [inline modules/aws/vpc/sg-etcd.tf]: oh neat. rebased/pushed this fix. Thanks!

**Comment by s-urbaniak** [comment]
@hhoover The AWS smoke test reveals:
```

* module.vpc.aws_security_group.etcd: At column 35, line 1: join: argument 1 should be type list, got type string in:

${var.external_sg_master == "" ?  join(" ", aws_security_group.master.id) : var.external_sg_master }

```

Do you mind to have a look? I believe the `join` is not necessary.

**Comment by hhoover** [comment]
removing the join results in computation errors again. Putting back the extra secgroup rule resources. 😞

**Comment by metral** [comment]
You're missing the `*` in the join - it needs to be: 

`${var.external_sg_master == "" ?  join(" ", aws_security_group.master.*.id) : var.external_sg_master }`

Test locally before pushing, but I'll defer to @s-urbaniak :)

**Comment by jasmingacic** [comment]
Is there anything else that needs to be done here? 
The PR seems to be addressing all that is required.

**Comment by hhoover** [comment]
Not that I know of, although it looks like it sat so long it needs another rebase. :-/

**Comment by s-urbaniak** [comment]
@hhoover @jasmingacic argh, agreed, sorry that this fell under the radar. if you don't mind, please rebase, so we can rekick the CI. /cc @alexsomesan

**Comment by jasmingacic** [comment]
I rebased the branch

**Comment by s-urbaniak** [comment]
@jasmingacic CI says
```
outdated docs (run 'make examples docs' to fix)
```

**Comment by s-urbaniak** [comment]
nevermind, that was an outdated CI build.

**Comment by s-urbaniak** [comment]
argh, it's playing tricks with me ;-) sorry for the fuzz, the above CI failure still is valid.

**Comment by jasmingacic** [comment]
I tried but I get this
```$ make examples docs
Using build directory [/Volumes/Data/go/src/github.com/hhoover/tectonic-installer/build/demo]
Error loading CLI configuration: 

open /Volumes/Data/go/src/github.com/hhoover/tectonic-installer/.terraformrc: no such file or directory
Makefile:96: *** "terraform-examples revision >= 83d7ad6 is required (https://github.com/s-urbaniak/terraform-examples)".  Stop.
```

**Comment by s-urbaniak** [comment]
@jasmingacic can you `go get -u github.com/s-urbaniak/terraform-examples` to update it to the newest version?

**Comment by jasmingacic** [comment]
Pushed the changes

**Comment by s-urbaniak** [comment]
ok to test

File: modules/aws/vpc/sg-etcd.tf [hunk 5]
```diff
@@ -48,3 +48,49 @@ resource "aws_security_group" "etcd" {
     self      = true
   }
 }
+
+# resource "aws_security_group_rule" "etcd_ssh_from_master" {
```
Comment (s-urbaniak) [inline modules/aws/vpc/sg-etcd.tf]: can we remove this whole commented block?

**Comment by s-urbaniak** [comment]
@jasmingacic this PR includes a merge commit https://github.com/coreos/tectonic-installer/pull/1389/commits/8d542b34c63acb47a27adcda3ba5f0366b2b6d89 which shouldn't be necessary. do you mind to remove it?

File: modules/aws/vpc/sg-etcd.tf (hunk 5 above)
Comment (jasmingacic) [reply modules/aws/vpc/sg-etcd.tf]: done

**Comment by jasmingacic** [comment]
@s-urbaniak do you want me to remove the commit from history or to remove what was added with the commit?

**Comment by jasmingacic** [comment]
I've been looking at [this](https://github.com/coreos/tectonic-installer/pull/1389/commits/8d542b34c63acb47a27adcda3ba5f0366b2b6d89) commit if it wasn't merged the PR would have conflicts.

**Comment by s-urbaniak** [comment]
@jasmingacic the commit "Merge remote-tracking branch 'upstream/master' into issue1163" seems unnecessary to me, no?

**Comment by jasmingacic** [comment]
ok to test

**Comment by jasmingacic** [comment]
I just removed the requested commit but it seems now the branch is not in sync with the upstream master.

**Comment by jasmingacic** [comment]
I rebased it but I was told that I should remove the merge commit.
What should I do next?

**Comment by s-urbaniak** [comment]
@jasmingacic do you mind to rebase it and resolve the conflicts during the rebase without adding a merge commit? I guess that is what I am actually asking for, sorry if this wasn't clear enough.

**Comment by jasmingacic** [comment]
Just to clarify before I proceed. What I usually do when I need to sync a fork is:

```
git fetch upstream/master
git merge upstream master
```

Then I commit the changes and push it to corresponding origin:[branch].

Is this what you want me to do or are we talking about something else?

**Comment by s-urbaniak** [comment]
@jasmingacic just a small tweak to your workflow:
```
git fetch upstream/master
git rebase upstream master
```

also see https://www.atlassian.com/git/tutorials/merging-vs-rebasing

**Comment by jasmingacic** [comment]
@s-urbaniak Done 

I didn't do `make examples docs` do you think I need to do that too?

**Comment by jasmingacic** [comment]
I just did it just in case

**Comment by mrwacky42** [comment]
We are evaluating Tectonic, and this is a problem we are having as well.

A different solution than proposed here would be to allow me to specify source addresses that are not `0.0.0.0/0` for the automatically created security groups.  
Alternatively, factor out the security group creation into another module that I can (optionally) specify trusted `cidr_blocks`.

Either way, I would not have to interpret what this statement means in order to bring up a Tectonic cluster.
> These groups must have proper ingress/egress rules for installation to be successful.

Thanks

**Comment by cpanato** [comment]
error on ci
```

Error loading configuration: Error loading /home/ubuntu/workspace/tectonic-installer_PR-1389-IX5DLHDAG2GYZVXI7UXD7SMBYJI7ALUNBLZF6N4XNTA42KZLAKGA@3/platforms/aws/main.tf: Error reading config for workers: At 218:22: unknown slice type: *ast.LiteralType
Makefile:39: recipe for target 'terraform-init' failed
make: *** [terraform-init] Error 1

```

**Comment by jasmingacic** [comment]
@s-urbaniak do you have an idea what might be happening here. We didn't have this error before.

**Comment by jasmingacic** [comment]
I think I found the issue but upon deeper look into the platforms/aws/main.tf I see two 
```
module "masters"{
``` 

at line 104 and 143

I'm not sure what is going on.

**Comment by s-urbaniak** [comment]
@jasmingacic that is most likely a merge conflict, as the master branch changed a lot since the initial PR. When back I can help with merging it, essentially all changes introduced here need to be merged with the current upstream master module declaration https://github.com/coreos/tectonic-installer/blob/master/platforms/aws/main.tf#L98.

**Comment by jasmingacic** [comment]
@s-urbaniak That would make sense if it was any other commit after the first one. The first commit contains this change. See below:
image

**Comment by jasmingacic** [comment]
@s-urbaniak I merged the best I could. Can you please check what I have here?

**Comment by jasmingacic** [comment]
ok to test

**Comment by jasmingacic** [comment]
ok to test

**Comment by jasmingacic** [comment]
ok to test

**Comment by jasmingacic** [comment]
ok to test

**Comment by cpanato** [comment]
Failures here is the Azure VPN, which is failing and we deactivated for now and the other failed to destroy.
I guess we can move on. @s-urbaniak @jasmingacic

**Comment by alexsomesan** [comment]
I share @s-urbaniak's POV entirely.
Adding `tectonic_aws_external_*_sg_id` when we already have an `tectonic_aws_*_extra_sg_ids` is really not good UX and will only fuel confusion (and associated support burden).

I support the idea of introducing a flag to toggle the creation of default SGs, in addition to the optionally supplied extra ones.

File: platforms/aws/main.tf [hunk 6]
```diff
@@ -151,6 +153,15 @@ module "ignition_workers" {
...
 
+  instance_count  = "${var.tectonic_worker_count}"
+  ec2_type        = "${var.tectonic_aws_worker_ec2_type}"
+  cluster_name    = "${var.tectonic_cluster_name}"
+  worker_iam_role = "${var.tectonic_aws_worker_iam_role_name}"
+
+  vpc_id                       = "${module.vpc.vpc_id}"
+  subnet_ids                   = ["${module.vpc.worker_subnet_ids}"]
+  sg_ids                       = ["${split(",", var.tectonic_aws_external_worker_sg_id == "" ? join(",", var.tectonic_aws_worker_extra_sg_ids, list(module.vpc.worker_sg_id)) : join(",", var.tectonic_aws_worker_extra_sg_ids, list(var.tectonic_aws_external_worker_sg_id)))}"]
+  ssh_key                      = "${var.tectonic_aws_ssh_key}"
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply platforms/aws/main.tf]: @s-urbaniak I'm not sure I understand. You want me to separate these four variables with a new line and then remove duplicates on lines 177 - 180?

[the remark this answers is not in the document]
Comment (s-urbaniak) [reply platforms/aws/main.tf]: @jasmingacic the whole block shouldn't contain duplicate entries yes. after a quick iteration I believe it should look like so:
```
  autoscaling_group_extra_tags = "${var.tectonic_autoscaling_group_extra_tags}"
  cl_channel                   = "${var.tectonic_cl_channel}"
  cluster_id                   = "${module.tectonic.cluster_id}"
  cluster_name                 = "${var.tectonic_cluster_name}"
  ec2_type                     = "${var.tectonic_aws_worker_ec2_type}"
  extra_tags                   = "${var.tectonic_aws_extra_tags}"
  instance_count               = "${var.tectonic_worker_count}"
  root_volume_iops             = "${var.tectonic_aws_worker_root_volume_iops}"
  root_volume_size             = "${var.tectonic_aws_worker_root_volume_size}"
  root_volume_type             = "${var.tectonic_aws_worker_root_volume_type}"
  sg_ids                       = ["${split(",", var.tectonic_aws_external_worker_sg_id == "" ? join(",", var.tectonic_aws_worker_extra_sg_ids, list(module.vpc.worker_sg_id)) : join(",", var.tectonic_aws_worker_extra_sg_ids, list(var.tectonic_aws_external_worker_sg_id)))}"]
  ssh_key                      = "${var.tectonic_aws_ssh_key}"
  subnet_ids                   = ["${module.vpc.worker_subnet_ids}"]
  vpc_id                       = "${module.vpc.vpc_id}"
  worker_iam_role              = "${var.tectonic_aws_worker_iam_role_name}"
```

[the remark this answers is not in the document]
Comment (jasmingacic) [reply platforms/aws/main.tf]: What about load_balancers 
```
  load_balancers               = ["${var.tectonic_aws_worker_load_balancers}"]
```

Should it go into the same block?

**Comment by jasmingacic** [comment]
How about we merge this before it goes out of sync with the master branch. 
I see that tests are failing, can someone pull me the logs why is that. Thanks

**Comment by s-urbaniak** [comment]
@jasmingacic master is effectively in code freeze now, so not too many things will happen. I rekicked CI since it was flaking. Having said that I propose to tackle the comments https://github.com/coreos/tectonic-installer/pull/1389#issuecomment-330236640 in this PR. Sorry for that, but I believe that is the better solution.

Having said that I have a task that interleaves with this functionality so I can help to implement the above proposal in the next few days.

**Comment by alexsomesan** [comment]
retest this please

**Comment by alexsomesan** [comment]
retest this please

**Comment by alexsomesan** [comment]
closing this in favor of #2029

**Pull request closed without merging**